src/ngraph/ngraph_graph.cc

```diff
@@ -262,96 +269,148 @@ std::vector<NodePtr> FindSubgraph(const Graph& graph, NodePtr node,
   // remove nodes on broken loops
   auto outNodes = RemoveBroken(node, subgraph_nodes);
 
-  // Ensure the graph is single output
-  outNodes = PruneSubgraphOutputs(graph, node, outNodes);
-
   return outNodes;
 }
 
+bool IdentifyOneSubgraph(Graph* graph, const std::function<bool(NodePtr)>& func,
+                         int current_subgraph_num, NodePtr n) {
+  bool found_subgraph = false;
+  if (n->subgraph_ == 0) {
+    // select nodes in the a subgraph starting here and going up the graph
+    auto subgraph_nodes = FindSubgraph(*graph, n, func);
+
+    // if we found a significantly large subgraph, label it
+    if (subgraph_nodes.size() > 0) {
+      for (auto node : subgraph_nodes) {
+        node->subgraph_ = current_subgraph_num;
+      }
+      CollapseSubgraph(graph, current_subgraph_num);
+      found_subgraph = true;
+    }
+  }
+  return found_subgraph;
+}
+
 // function to identify and label connected ngraph ops as subgraphs
-void IdentifySubgraphs(const Graph& graph,
-                       const std::function<bool(NodePtr)>& func) {
+void IdentifySubgraphs(Graph* graph, const std::function<bool(NodePtr)>& func) {
   int sg = 1;
-  // loop over the nodes from the back
-  for (auto n = graph.nodes_.rbegin(); n != graph.nodes_.rend(); ++n) {
-    if ((*n)->subgraph_ == 0) {
-      // select nodes in the a subgraph starting here and going up the graph
-      auto subgraph_nodes = FindSubgraph(graph, *n, func);
-
-      // if we found a significantly large subgraph, label it
-      if (subgraph_nodes.size() > 0) {
-        for (auto node : subgraph_nodes) {
-          node->subgraph_ = sg;
-        }
-        for (auto node : subgraph_nodes)
-          for (auto i : node->inputs_) {
-            if (i->subgraph_ != sg) i->subgraph_ = -1;
-          }
 
+  // collapse graphs from the outputs
+  for (size_t i = 0; i < graph->num_outputs_; ++i) {
+    bool found_subgraph = false;
+    found_subgraph = IdentifyOneSubgraph(graph, func, sg, graph->outputs_[i]);
+    if (found_subgraph) {
+      sg += 1;
+    }
+  }
+
+  // loop over the nodes from the back and find any that aren't in subgraphs
+  while (true) {
+    bool found_subgraph = false;
+    for (auto n = graph->nodes_.rbegin(); n != graph->nodes_.rend(); ++n) {
+      found_subgraph = IdentifyOneSubgraph(graph, func, sg, *n);
+      if (found_subgraph) {
         sg += 1;
+        break;
       }
     }
+    if (found_subgraph) {
+      continue;
+    } else {
+      break;
+    }
   }
 }
 
 // Function to collapse the intermediary graph into a graph
 // with subgraphs for nodes
-void CollapseSubgraphs(Graph* graph) {
+void CollapseSubgraph(Graph* graph, int subgraph_num) {
```

Comment:
Is `graph` intentionally a `*` rather than a `&`?

In my own coding, I usually use `*` to hint that the identifier is sometimes permitted to be null.

Reply:
MXNet cpplint doesn't allow you to pass a non-const reference. I want to modify it, so it's a pointer.